### ISSUE TITLE
Massive GB performance regressions between 6.2 and 6.3

### ISSUE THREAD

**Issue by jlmaccal** [pandegroup/openmm#1502]
We have largely been using older versions of OpenMM with MELD, so we didn't catch this sooner.  There were **MASSIVE** performance decreases for GB simulations between 6.2 and 6.3, that persist up to the current master.

Running the test system below results in 200 ns / day with OpenMM 6.2. Running the same system with 6.3 or later results in **31 ns / day!**

The test system is here:
<https://gist.github.com/ca6b08a307800f80986661075c82c432>

Results are from ubuntu 14.04 with cuda 7.5.
NVIDIA GTX 780
CUDA 7.5

We expect similar results with Tesla 2070. I don't yet have access to newer GPUs for testing.

I've spent the last few hours with `git bisect` and have identified the following culprits. Some appear to be related to changes @peastman made to implment triclinic boxes. Some appear specific to changes @swails made to gbNeck models.

Commit | Description | Speed change (ns/day)
-----------|-----------------|--------------------------------
61458a5 | Beginning of CUDA support for triclinic boxes | 200 to 158
6eb302f | Continuing to implement CUDA support for triclinic boxes | 158 to 136
dafd0f9 | GBn now uses Discrete2DFunctions | 136 to 141 (this is slightly faster)
4c450bf | Move to using a continuous 2D function. | 141 to 31

**Comment by jlmaccal** [comment]
I've done some profiling on the last commit. 

The bottleneck appears to be memory latency when using tabulated functions, as in the gbNeck models. The kernel needs to determine the values for `m0` and `d0`. It actually does this twice with the radii swapped the second time. There is obviously a difference in the number of arithmetic operations between the continuous and discrete lookups, but the continuous version is clearly bound by memory latency, not compute.

With the continuous lookup, the kernel loads 64 floats per iteration from global memory. With the discrete table, it loads 4 floats per iteration. There is no relationship between the thread index and the address of these loads. Furthermore, it does the lookups in the order `m0`, `d0`, `m0`, then `d0`, with the radii swapped in the second two lookups.

So, there is a huge difference in performance between the discrete and continuous versions, with a corresponding 16-fold difference in global memory lookups. I'm not sure if this is simply due to the higher volume of global memory accesses, or if caching plays a role as well.

It's not obvious to me what the best way is to fix this It seems like copying the lookup table(s) to shared memory ought to give an enormous speedup, but this could lower occupancy. If these were hard-coded kernels, I would just try this. But since they are generated code, I'm not sure exactly how to make these changes.

This is probably a performance issue everywhere that the tabulated functions are used, but it's a more serious issue here because they are being used in the inner loop of a non-bonded function.

**Comment by peastman** [comment]
@swails can give more insight on why this change was made.  My vague recollection is that the old method (looking up fixed values rather than interpolating) was inconsistent with how Amber implements it.  But I'm not sure how important that really is, and it certainly has a big effect on performance.

Short of switching back to the old behavior, I'm not sure what can be done about it.  2D interpolation requires a lot more data than a simple lookup.  Shared memory is much less important on modern GPUs than it used to be.  They often use exactly the same memory for both shared and cache, so if there's too much data to fit in cache, there also is too much data to hold in shared memory (at least without dramatically reducing occupancy).  In the case of GBn, the two tables have a total of 882 values.  Each one gets turned into a 2D spline patch with 16 coefficients, each represented as a 4 byte float, so that's 56K: way too much to copy to shared memory.  If it weren't, this problem wouldn't exist because the cache would already be buffering it effectively.

**Comment by jlmaccal** [comment]
I thought the tables were 20x20? Where does 882 come from? In any case, I forgot the factor of 4 from floats to bytes, so I see your point :)

But in this case, I think we can improve the implementation. We obviously don't need the spline interpolation on the GPU, as there are a fixed number of (constant) radii. We can precompute these and then lookup in that much smaller discrete table on the GPU. I can implement this.

Can you look into the slowdowns related to triclinic boxes? These seem to cut 40% of performance, even for non-periodic systems.

**Comment by peastman** [comment]
The tables are 21x21, and there's two of them.  21*21*2=882.

**Comment by jlmaccal** [comment]
Gotcha! I thought you meant each. In any case, I was missing a factor of 4.

**Comment by jchodera** [comment]
Given that there are already huge differences in the way that OpenMM and AMBER implement the SASA term, are we really concerned by the differences in the GBn implementation that led to such a huge performance change?  I wonder if there should at least be a flag to select which behavior (fast and approximate, slow and more faithful) is desired.

**Comment by jlmaccal** [comment]
I think we can fix this to have the correct behavior and be fast. Working on it.

**Comment by swails** [comment]
What Amber does is a linear interpolation between the nearby grid points. When the radii are exactly on the grid points, the discrete function gets the right answer. 

But when the radii are not on the grid points, the continuous function does a better job. Also, when you read the GBneck paper by Mongan, the continuous 2d function is far closer to the true intent of the model. 

I didn't realize there was a pronounced performance effect when I made the PR. But you can go back to the PR to see the discussion regarding why the change was made (and ultimately merged).

**Comment by jlmaccal** [comment]
This should be partially fixed by #1504.

I now get 150 ns/day on the test system, so it's actually slightly faster, while now matching Amber's behaviour more closely.

There is still a ~25% slowdown due to support for triclinic boxes. It's not clear to me why this should slow down simulations in non-periodic systems.

**Comment by jlmaccal** [comment]
> Given that there are already huge differences in the way that OpenMM and AMBER implement the SASA term

@jchodera What are the differences?

**Comment by jchodera** [comment]
AMBER uses a [linear combination of pairwise overlaps (LCPO) scheme](http://onlinelibrary.wiley.com/doi/10.1002/(SICI)1096-987X(19990130)20:2%3C217::AID-JCC4%3E3.0.CO;2-A/full?2-A/full), while OpenMM uses a totally different [function from Tinker inspired by the ACE model](http://docs.openmm.org/7.0.0/userguide/theory.html#surface-area-term). They give totally different numerical results, and have little relation to each other in derivation.

**Comment by peastman** [comment]
I've been working a lot to figure out where the remaining slowdown comes from.  I think I now partly understand it.

First of all, I don't think it has anything to do with triclinic boxes.  I checked the revisions you mentioned, and I don't observe any change in performance from them.  There are only two revision where I see any change: https://github.com/pandegroup/openmm/commit/dafd0f9754b769cd82b2e236906c6515b134d938, which moved from Discrete1DFunction to Discrete2DFunction, and then https://github.com/pandegroup/openmm/commit/4c450bf5ffd1af1846aefe92ef47e03796428b50 which moved to Continuous2DFunction.  For everything prior to https://github.com/pandegroup/openmm/commit/dafd0f9754b769cd82b2e236906c6515b134d938, I get 370 ns/day on a GTX 980.   In that revision it decreased to 167 ns/day.  The second revision then brought it down further to only about 60 ns/day.

The second one makes sense: continuous functions involve a lot more memory access.  But why would changing from a 1D to a 2D function have such a big effect?  The generated code is very very similar.  As far as I can tell, the difference comes from rounding.

For reference, here is how the original version calculated its index for the Discrete1DFunction:

```Python
"index = (radius2*200-20)*21 + (radius1*200-20);"
```

And here is how the Discrete2DFunction version computed its indices:

```Python
"index1=radius1*200-20; index2=radius2*200-20;"
```

They look very similar, but the first one actually produced incorrect results, which is why I replaced it.  The problem is that `(radius2*200-20)` isn't necessarily an integer.  So when the fractional part of it gets multiplied by 21 and added to `(radius1*200-20)`, it can lead to the wrong value being computed for `index`.  Discrete2DFunction doesn't let that happen.  It rounds each index to the nearest integer to ensure the result is correct.

Suppose we stick with Discrete1DFunction but force the first index to be an integer:

```Python
"index = floor(radius2*200-20)*21 + (radius1*200-20);"
```

With this version I get only 146 ns/day, even slower than the Discrete2DFunction version!  For some reason, `floor()` appears to be a really expensive function.  I don't know why.

So what can we do about it?  Well, in your revised version we don't need to compute the indices individually.  There's a per-particle parameter that already gives the integer index.  So you can use a Discrete1DFunction, compute the index in the table yourself, and *not* include a `floor()` function because you know it's already an integer.  Give that a try and see how it works.

**Comment by jlmaccal** [comment]
@peastman Sorry for the delay in responding. I was travelling for the last two weeks.

You are indeed correct about the triclinic box not having any effect. I'm not sure what happened, as I did get those timings while running `git bisect`, but now when I go back and test, I get the same results as you. I wonder if something wasn't properly rebuilt when I was testing the intermediate commits?

In any case, I'll check your suggestion about avoiding the floor function.

**Comment by jlmaccal** [comment]
@peastman OK, so there is some issue with caching that's making this very confusing.

If I checkout `6eb302f`, build from scratch, and run my benchmark, I get 200 ns / day. If I then checkout `dafd0f9~` (ie. the parent of dafd0f9), build from scratch, then I still get 200 ns/day. However, if I delete the cached kernels with `rm /tmp/*_64` and re-run the benchmark, then it drops to 135 ns / day. If I then checkout `6eb302f`, build from scratch, and run the benchmark, it stays at 135 ns/day. If I then `rm /tmp/*_64`, then it jumps back up to 200 ns / day.

So, there's something strange going on with caching, although I'm not yet sure what or how to trigger it.

I believe these cached kernels are leading to confusing benchmark results.

**Comment by jlmaccal** [comment]
@peastman 

OK, I've tracked down the remaining speed difference, which is about 25% on my machine for the test case linked above. The code generated by the cuda compiler plugin is slower than that generated by nvcc. I have no idea why this is, but I've been able to reproduce this on two different GTX 780 machines.

This is a bit tricky to test, as you need to delete the cached kernels between tests. If you don't it will used the cached copies as the files are named according to hash of the code to be compiled, which doesn't change. But the runtime compiler and nvcc do produce different ptx files. This is also why I was getting weird results using `git bisect`.

As I said, I'm not sure why there is a difference. The parameters passed seem to be equivalent, so I suspect that there is a different default for one of the options depending on if the runtime compiler or nvcc is used. I'll keep digging.

**Comment by peastman** [comment]
Wow, that's really interesting.  Presumably it means whatever compiler they've built into the runtime compilation library isn't as good as whatever external compiler nvcc was using?

**Comment by jlmaccal** [comment]
I can't see any flags that would lead to this difference. The runtime compiler and nvcc generate different ptx code with very different performance.

Kernel | runtime compiler (ms per kernel invocation) | nvcc (ms per kernel invocation)
----------|-----------------------------------------------------------|-------------------------------------------
computeNonbonded | 105 | 68
computeN2Value | 59 | 31
computeN2Energy | 27 | 25

**Comment by jlmaccal** [comment]
> Wow, that's really interesting. Presumably it means whatever compiler they've built into the runtime compilation library isn't as good as whatever external compiler nvcc was using?

That seems to be the most likely explanation at this point.

**Comment by peastman** [comment]
What external compiler is nvcc using on your computer?  And what exactly does "ms per kernel invocation" mean in the table above?  How did you measure it?

**Comment by jlmaccal** [comment]
> What external compiler is nvcc using on your computer? 

No idea. It's ubuntu 14.04, so I assume gcc 4.8.4.

> And what exactly does "ms per kernel invocation" mean in the table above? How did you measure it?

As it says on the tin, it's the time taken for each call to that kernel in the test case linked above. Measured with `nvvp`.

**Comment by jlmaccal** [comment]
Actually, I lied... I believe its the total for all invocations of each kernel in the test case above.

**Comment by jlmaccal** [comment]
I don't think there is anything that can be done to fix this, as it appears to be a difference with the runtime's internal compiler.

If this issue can be reproduced on another machine, I think the best course of action is to add an environment variable to specify if the runtime plugin should be used, and to possibly disable it by default. On my GB test system the performance difference is close to 30 percent.

Should I create a separate issue for this?

**Comment by peastman** [comment]
Sounds like a good idea.

For what it's worth, I've never seen that problem.  Possibly it's because you're using an older GPU?

**Comment by jlmaccal** [comment]
> For what it's worth, I've never seen that problem. Possibly it's because you're using an older GPU?

I suppose that's possible, but I don't see why there would be a slowdown for one architecture, but not another, unless their internal compiler is somehow worse with older hardware.

Do you want to test things on a more recent card? I've been using `e84f072` and `a0bdc69`. But don't forget to wipe the cached ptx files, as it won't recompile them otherwise. Perhaps if it works fine on newer cards we can make the use of the runtime compiler depend on the architecture? Perhaps with an environment variable to force one or the other?

Along those lines, it might be a good idea to include the git hash in the temporary file names, so that any code changes always result in recompilation. Most of the time, this would be unnecessary, but it would have caught this case. Diagnosing this made me question my sanity for a while...

**Comment by jlmaccal** [comment]
@peastman , I've tested your suggestions for indexing, but on my hardware and with `nvcc`, using `Discrete2DFunction` directly seems to give better performance than `Discrete1DFunction` with an explicitly calculated index.

**Comment by peastman** [comment]
Could you post the code so I can see how you did it?

**Comment by jlmaccal** [comment]
I guess I can reproduce it, but I threw it away… it’s basically something like this. I replaced the 2D functions with 1D functions and modified the expression as below.

```
...
m0 = getm0(index);
d0 = getd0(index);
index = radindex2*{n_rad_indices} + radindex1;
…
```

This is then passed through `format` to replace `n_rad_indices` with the appropriate value.

This seems to be ~5% slower on my machine.

**Comment by peastman** [comment]
I can reproduce the difference from the runtime compiler.  The difference I'm seeing is a bit smaller (around 10%), but I've reproduced it on both a K80 (Kepler) and a GTX 980 (Maxwell).  Sigh.

For the moment, an easy workaround is to delete libOpenMMCudaCompiler from your plugins directory, and disable OPENMM_BUILD_CUDA_COMPILER_PLUGIN in CMake.  But how do we want to deal with this?  Right now, if the runtime compiler is available it uses it, but that means it prefers the slower compiler over the faster one.  I could make it check whether nvcc is present and if so, use it?  There's still a possibility that nvcc could fail due to not having an external C compiler available.  But presumably people who install the CUDA toolkit are likely to already have a C compiler.

I'll also submit a bug report to Nvidia.

**Comment by jlmaccal** [comment]
I think your suggestion would work. Another possibility is to disable the runtime compiler unless it's specifically enabled by an environment variable.

**Comment by jlmaccal** [comment]
@peastman 

Did you decide on a course of action here? In my opinion, users should get the best performance without having to delete libraries. My vote would be to disable until the performance issues are sorted.

**Comment by peastman** [comment]
Nvidia has reproduced the problem and they're looking into it.  I'm waiting to see what they find.  But assuming there's no better solution by the time we're ready for the next release, I'll make it default to using nvcc if it's available.

**Comment by peastman** [comment]
Between #1504 and #1549, I believe this is now fixed.  @jlmaccal are you ok with this issue being closed?

**Comment by jlmaccal** [comment]
Yes, I think we're good to close it.

**Comment by peastman** [comment]
I just got a notice from Nvidia that they've fixed the problem with the runtime compiler, and the fix will be in CUDA 8.0.
